store/src/java/com/zimbra/cs/db/DbMailItem.java

```diff
@@ -2778,6 +2779,54 @@ public static UnderlyingData getByHash(Mailbox mbox, String hash) throws Service
         }
     }
 
+    public static int getModifiedItemsCount(Mailbox mbox, MailItem.Type type, long lastSync,
+            int sinceDate, Set<Integer> visible)
+    throws ServiceException {
+        if (Mailbox.isCachedType(type)) {
+            throw ServiceException.INVALID_REQUEST("folders and tags must be retrieved from cache", null);
+        }
+        DbConnection conn = mbox.getOperationConnection();
+        PreparedStatement stmt = null;
+        try {
+            String typeConstraint = type == MailItem.Type.UNKNOWN ? "type NOT IN " + NON_SYNCABLE_TYPES : typeIn(type);
```

Comment:
There are a lot of intermediate Strings being created here to create the statement.  How about sticking with `StringBuilder` syntax and doing something like:

    buf.append("SELECT COUNT(id)")
        .append(" FROM ")
        .append(getMailItemTableName(mbox))
        .append(" WHERE ")
        .append(IN_THIS_MAILBOX_AND)
        .append("mod_metadata > ? AND ");
        .append(sinceDate > 0 ? "date > ? AND " : "")
        .append(MailItem.Type.UNKNOWN ? "type NOT IN " + NON_SYNCABLE_TYPES : typeIn(type))
    if (visible != null && visible.size() > 0) {
        String folderString = visible.stream().map(i -> i.toString()).collect(Collectors.joining(", "));
        buf.append(" AND folder_id IN ( ").append(folderString).append(" )");
    }

Reply:
Done